src/Cleanup.php

```diff
@@ -118,123 +133,117 @@ public function preMigration(string $mainRoleName): void
-                    );
-                    $this->destinationConnection->useRole($role['granted_by']);
-                }
+                $dataToRemove = $this->getDataToRemove($this->destinationConnection, $roleName, $mainRoleName);
+            } else {
+                $this->logger->info(sprintf('Database %s exists, getting ownership role', $database));
+                $databaseRole = $this->destinationConnection->getOwnershipRoleOnDatabase($database);
+                $dataToRemove = $this->getDataToRemove($this->destinationConnection, $databaseRole, $mainRoleName);
+            }
 
-                /** @var FutureGrantToRole[] $futureGrants */
+            // First revoke all future grants from roles
+            foreach ($dataToRemove['ROLE'] as $role) {
+                $this->destinationConnection->useRole($role['granted_by']);
                 $futureGrants = array_map(
                     fn(array $v) => FutureGrantToRole::fromArray($v),
                     $this->destinationConnection->fetchAll(sprintf(
                         'SHOW FUTURE GRANTS TO ROLE %s',
-                        $role['name']
+                        Helper::quoteIdentifier($role['name'])
                     ))
                 );
+
+                if ($futureGrants) {
+                    [$sqls, $currentRole] = $this->switchRole($role['granted_by'], $mainRoleName, $sqls, $currentRole);
+                }
                 foreach ($futureGrants as $futureGrant) {
                     $sqls[] = sprintf(
                         'REVOKE %s ON FUTURE TABLES IN SCHEMA %s FROM ROLE %s;',
                         $futureGrant->getPrivilege(),
                         $futureGrant->getName(),
-                        Helper::quoteIdentifier($futureGrant->getGranteeName()),
+                        Helper::quoteIdentifier($futureGrant->getGranteeName())
                     );
                 }
-                $sqls[] = sprintf('DROP ROLE IF EXISTS %s;', Helper::quoteIdentifier($role['name']));
             }
 
-            $projectUser = $this->getProjectUser($databaseRole);
-
-            if ($projectUser->getGrantedBy() !== $currentRole) {
-                $currentRole = $projectUser->getGrantedBy();
-                if ($currentRole === $mainRoleName && !$mainRoleExistsOnTargetUser) {
-                    $sqls[] = sprintf(
-                        'USE ROLE %s;',
-                        Helper::quoteIdentifier($this->config->getTargetSnowflakeRole())
-                    );
-                    $sqls[] = sprintf(
-                        'GRANT ROLE %s TO USER %s;',
-                        Helper::quoteIdentifier($currentRole),
-                        Helper::quoteIdentifier($this->config->getTargetSnowflakeUser())
-                    );
-                }
-                $sqls[] = sprintf('USE ROLE %s;', Helper::quoteIdentifier($currentRole));
+            // Drop users owned by roles
+            if (!empty($dataToRemove['USER'])) {
```

Comment:
Ty `'USER'` a `'ROLE'` stringy jsi asi chtěl i tady nahradit `self::USER`, resp. `self::ROLE`, žejo 🙂

Reply:
jo určitě :)